dex/testing/xmr/monero_functions.inc

```diff
@@ -0,0 +1,390 @@
+	local amount="${2:?input an amount of monero to send in atomic units 1e12}"
+	local address="${3:?input a valid destination wallet address}"
+	local unlock_time=${4:-0} # unlock after n blocks
+	destinations="[{\"amount\":${amount},\"address\":\"${address}\"}]"
+    other_params="\"account_index\":0,\"subaddr_indices\":[0],\"priority\":0,\"unlock_time\":${unlock_time},\"do_not_relay\":true,\"get_tx_key\":true,\"get_tx_hex\":true,\"get_tx_metadata\":true"
+	params="{\"destinations\":[{\"amount\":${amount},\"address\":\"${address}\"}], "${other_params}" }"
+	monero-rpc-request "${port}" "transfer" "${params}"
+}
+
+# sign a prepared unsigned txset -- offline manual experimental
+# 
+# Note: can only be done by an offline cold wallet that has never been "online"
+# use the monero-wallet-cli interactive tool instead in a manual process with
+# an unconnected charlie_view wallet which holds the inputs but not any keys.
+sign_transfer() {
+	local port="${1:?input a valid wallet rpc port for a cold wallet}"
+	local unsigned_txset="${2:?input an unsigned tx set}"
+	params="{\"unsigned_txset\":\"${unsigned_txset}\"}"
+	monero-rpc-request "${port}" "sign_transfer" "${params}"
+}
+
+cmd_help() {
+	echo "Commands Help:"
+	echo "--------------"
+	echo "alpha_get_transactions"
+	echo "- get transaction details for one or more txid"
+	echo "- inputs:"
+	echo "  - tx_hashes - hash1,hash2,hash3,..."
+	echo ""
+	echo "alpha_get_transactions_details"
+	echo "- get transaction development details from monerod including tx lock time"
+	echo "- inputs:"
+	echo "  - tx_hashes - hash1,hash2,hash3,..."
+	echo ""
+	echo "alpha_info"
+	echo "- get running daemon details - height, etc."
+	echo "- inputs: None"
+	echo ""
+	echo "alpha_sendrawtransaction"
+	echo "- broadcast a previously built signed tx"
+	echo "- inputs:"
+	echo "  - tx_as_hex string - can be generated with charlie_build_tx or fred_build_tx"
+	echo ""
+	echo "alpha_transaction_pool"
+	echo "- get mempool details"
+	echo "- inputs: None"
+	echo ""
+	echo "mine-to-bill"
+	echo "- generate 1 or more blocks to bill wallet"
+	echo "- inputs:"
+	echo "  - num_blocks - defaults to 1"
+	echo ""
+	echo "bill_balance"
+	echo "- get bill wallet balance details"
+	echo "- inputs: None"
+	echo ""
+	echo "bill_refresh_wallet"
+	echo "- update bill's wallet from the daemon latest info"
+	echo "- inputs: None"
+	echo ""
+	echo "bill_transfer_to"
+	echo "- build, sign and broadcast a transaction from bill wallet to another address"
+	echo "- inputs:"
+	echo "  - amount in in atomic units 1e12 - e.g. 1230000000000 = 1.23 XMR"
+	echo "  - address - recipient primary address - account index 0, subaddr_indeces [0]"
+	echo ""
+	echo "charlie_balance"
+	echo "- get charlie wallet balance details"
+	echo "- inputs: None"
+	echo ""
+	echo "charlie_refresh_wallet"
+	echo "- update charlie's wallet from the daemon latest info"
+	echo "- inputs: None"
+	echo ""
```

Comment:
Is refresh the same as a rescan?

Reply:
No .. refresh is a way for the wallet to ask the daemon to update it with any new transactions that have arrived for it from mempool or already in a recent block.
[refresh](https://web.getmonero.org/resources/developer-guides/wallet-rpc.html#refresh)
There is also:
[rescan_blockchain](https://web.getmonero.org/resources/developer-guides/wallet-rpc.html#rescan_blockchain)
Re-scan the blockchain from scratch, losing any information which can not be recovered from the blockchain itself.
This includes destination addresses, tx secret keys, tx notes, etc.
[I never tried this]
[rescan_spent](https://web.getmonero.org/resources/developer-guides/wallet-rpc.html#rescan_spent)
Re-scan the blockchain for spent outputs.

Neither of the rescans return any info (except Json 2 `result` or `error`